metals/src/main/scala/scala/meta/internal/metals/debug/DebugProvider.scala

```diff
@@ -275,6 +276,51 @@ class DebugProvider(
     }
   }
 
+  private def resolveInFile(
+      buildTarget: BuildTargetIdentifier,
+      classes: TrieMap[
+        BuildTargetClasses.Symbol,
+        ScalaMainClass
+      ],
+      testClasses: TrieMap[
+        BuildTargetClasses.Symbol,
+        BuildTargetClasses.ClassName
+      ],
+      params: DebugDiscoveryParams
+  )(implicit ec: ExecutionContext) = {
+    val path = params.path.toAbsolutePath
+    semanticdbs
+      .textDocument(path)
+      .documentIncludingStale
+      .fold[Future[DebugSessionParams]] {
+        Future.failed(SemanticDbNotFoundException)
+      } { textDocument =>
+        lazy val tests = for {
+          symbolInfo <- textDocument.symbols
+          symbol = symbolInfo.symbol
+          testClass <- testClasses.get(symbol)
+        } yield testClass
+        val mains = for {
+          symbolInfo <- textDocument.symbols
+          symbol = symbolInfo.symbol
+          mainClass <- classes.get(symbol)
+        } yield mainClass
+        if (mains.nonEmpty) {
+          verifyMain(buildTarget, mains.toList, params)
+        } else if (tests.nonEmpty) {
+          Future(
+            new b.DebugSessionParams(
+              singletonList(buildTarget),
+              b.DebugSessionParamsDataKind.SCALA_TEST_SUITES,
+              tests.asJava.toJson
+            )
+          )
+        } else {
+          Future.failed(NoRunOptionException)
```

Comment:
_Total aside_, but on my TODO list to to refactor all this fails out. I'm not a huge fan of using failing futures for control flow. I think we can probably model all of these differently. One day 😆

Reply:
Are you talking about a dedicated error channel?